File: src/main/scala/org/training/spark/sql/RDD2DataframeReflection.scala
package org.training.spark.sql

import org.apache.spark.SparkConf
import org.apache.spark.sql.SparkSession

/**
  * Created by anderson on 17-9-16.
  */

case class Persons(name : String,age : Int)

object RDD2DataframeReflection {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("RDD2DataframeReflection")
    val spark = SparkSession
      .builder()
      .config(conf)
      .getOrCreate()

    import spark.implicits._
    // 先创建RDD, 然后调用toDF()方法转换为DataFrame, 调用toDF()方法需要导入隐式转换: import spark.implicits._
    val peopleDF = spark.sparkContext
      .textFile("hdfs://anderson-JD:9000/examples/people.txt")
      .map(_.split(","))
      .map(attributes => Persons(attributes(0), attributes(1).trim.toInt))
        .toDF()

    peopleDF.createOrReplaceTempView("people")

    val teenagersDF = spark.sql("SELECT name, age FROM people WHERE age BETWEEN 13 AND 19")


    teenagersDF.rdd.foreach(row => println(row.getString(0) + " " + row.getString(1)))
    teenagersDF.rdd.saveAsTextFile("hdfs://anderson-JD:9000/examples/RDD2DataframeProgrammactically")

    teenagersDF.map(teenager => "Name:" + teenager(0) + ", age:" + teenager(1)).show()
    implicit val mapEncoder = org.apache.spark.sql.Encoders.kryo[Map[String, Any]]
    teenagersDF.map(teenager => teenager.getValuesMap[Any](List("name", "age"))).collect()
  }
}

File: src/main/scala/org/training/spark/sql/RDD2DataFrameReflection.java
package org.training.spark.sql;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoder;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

/**
 * Created by anderson on 17-9-16.
 */
public class RDD2DataFrameReflection
{
    public static void main(String[] args)
    {
        SparkConf sparkConf = new SparkConf();
//        conf.setMaster("local");
        sparkConf.setAppName("RDD2DataFrameReflection");
        JavaSparkContext sparkContext = new JavaSparkContext(sparkConf);
        SparkSession spark = SparkSession
                .builder()
//                .appName("Java Spark SQL basic example")
//                .config("spark.some.config.option", "some-value")
                .config(sparkConf)
                .getOrCreate();

        JavaRDD<Person> personRDD = sparkContext.textFile("hdfs://anderson-JD:9000/examples/people.txt").map(new Function<String, Person>() {
            @Override
            public Person call(String line)
                    throws Exception
            {
                String[] strs = line.split(",");
                String name = strs[0];
                int age = Integer.parseInt(strs[1].trim());
                Person person = new Person(age, name);
                return person;
            }
        });
        // rdd -> dataFrame
        Dataset<Row> peopleDF = spark.createDataFrame(personRDD, Person.class);
        // Register the DataFrame as a temporary view
        peopleDF.createOrReplaceTempView("people");
        Dataset<Row> teenagersDF = spark.sql("SELECT name, age FROM people WHERE age BETWEEN 13 AND 19");

        Encoder<String> stringEncoder = Encoders.STRING();
        Dataset<String> teenageNamesByIndexDF = teenagersDF.map(
                (MapFunction<Row, String>) row -> "Name:" + row.getString(0) + ", age:" + row.getInt(1), stringEncoder);

        teenageNamesByIndexDF.show();
        // 保存成文件
//        teenageNamesByIndexDF.javaRDD().saveAsTextFile("hdfs://anderson-JD:9000/examples/result");
    }
}
